### PR TITLE
Add word-break attribute for text

### LINKED ISSUE #1164: Not-so-prefered word wrapping on Korean words
### Description

Here's a 2 dummy Korean sentences:

> 뉴질랜드 북섬 호크스베이 지역에 위치한 포랑아하우에 있는 언덕인 타우마타와카탕이항아코아우아우오타마테아투리푸카카피키마웅아호로누쿠포카이웨누아키타나타후는 줄여서 타우마타와카탕이항아코아우아우오타마테아포카이웨누아키타나타후이며, 더 줄여서 타우마타라고 불리기도 한다.

> 세종대왕의 맥북프로 던짐 사건은 역사서적인 조선왕조실록에 기록된 일화로, 15세기 조선시대 세종대왕이 새로 개발한 훈민정음(한글)의 초고를 작성하던 중, 문서 작성 중단에 대한 담당자에게 분노하여 맥북프로와 함께 그를 방으로 던진 사건입니다.

And here is the render result with #335 sample texts:

image

Typst break some words that should not broken:
- 세종대왕이 => 세종대왕/이
- 맥북프로와 => 맥북/프로와
- 타우마타와카탕이항아코아우아우오타마테아포카이웨누아키타나타후이며 => 타우마타와카탕이항아코/아우아우오타마테아포카이웨누아키타나타후이며

Korean generally does not prefer breaking words, but prefers line break before a word instead. Something like [`word-break: keep-all` option in CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/word-break) would be good?

Example of expected word break:

```
뉴질랜드 북섬 호크스베이 지역에 위치한 포랑아하우에 있는 언덕인
타우마타와카탕이항아코아우아우오타마테아투리푸카카피키마웅아호로누쿠포카이웨누아키타나타후는 줄여서
타우마타와카탕이항아코아우아우오타마테아포카이웨누아키타나타후이며, 더 줄여서 타우마타라고 불리기도
한다.

세종대왕의 맥북프로 던짐 사건은 역사서적인 조선왕조실록에 기록된 일화로, 15세기 조선시대
세종대왕이 새로 개발한 훈민정음(한글)의 초고를 작성하던 중, 문서 작성 중단에 대한 담당자에게 분노하여
맥북프로와 함께 그를 방으로 던진 사건입니다.
```

Or more narrow width:
```
뉴질랜드 북섬 호크스베이 지역에 위치한 포랑아하우에 있는 언덕인
타우마타와카탕이항아코아우아우오타마테아투리푸카카피키마웅아호로
누쿠포카이웨누아키타나타후는 줄여서
타우마타와카탕이항아코아우아우오타마테아포카이웨누아키타나타후이
며, 더 줄여서 타우마타라고 불리기도 한다.

세종대왕의 맥북프로 던짐 사건은 역사서적인 조선왕조실록에 기록된
일화로, 15세기 조선시대 세종대왕이 새로 개발한
훈민정음(한글)의 초고를 작성하던 중, 문서 작성 중단에 대한
담당자에게 분노하여 맥북프로와 함께 그를 방으로 던진 사건입니다.
```

Maybe related: #335 

### Reproduction URL

https://typst.app/project/rRVJfqU5x1LjoJOOtx0ZoU

### Browsers

Firefox

### OS

Linux

### PR DESCRIPTION
**Pull request by RanolP**
Fixes #1164. In the future, we'll need to implement [overflow-wrap](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-wrap) to cover some edge cases (we can see that the texts are clipped in the new test image).

### CODE REVIEW

File: crates/typst/src/text/mod.rs [hunk 1]
```diff
@@ -464,6 +464,21 @@ pub struct TextElem {
...
+    /// 줄 나눔 기준을 “글자” 또는 “어절” 단위로 한다.
+    ///
+    /// #set text(word-break: "keep-all")
+    /// 각 줄의 마지막에 한글이 올 때
+    /// 줄 나눔 기준을 “글자” 또는 “어절” 단위로 한다.
+    /// ```
+    #[default(WordBreak::Normal)]
+    #[ghost]
+    pub word_break: WordBreak,
+
```
Comment (PgBiel) [inline crates/typst/src/text/mod.rs:481]: I'd suggest starting with one or two examples in English, solely for consistency with the rest of the documentation. But keeping at least one CJK sample is also fine imo, however I don't know what the rules are here.

File: crates/typst/src/text/linebreak.rs [hunk 2]
```diff
@@ -42,3 +44,26 @@ impl Behave for Packed<LinebreakElem> {
         Behaviour::Destructive
     }
 }
+
+/// A word-break mode on text
```
Comment (PgBiel) [inline crates/typst/src/text/linebreak.rs:48]:
```suggestion
/// A word break mode on text.
```

File: crates/typst/src/layout/inline/linebreak.rs [hunk 3]
```diff
@@ -3,30 +3,36 @@ use icu_properties::LineBreak;
...
 use once_cell::sync::Lazy;
 
 use super::Preparation;
 use crate::syntax::link_prefix;
-use crate::text::{Lang, TextElem};
+use crate::text::{Lang, TextElem, WordBreak};
 
-/// The general line break segmenter.
-static SEGMENTER: Lazy<LineSegmenter> = Lazy::new(|| {
+/// Create a general line break segmenter with word-break option.
```
Comment (PgBiel) [inline crates/typst/src/layout/inline/linebreak.rs:13]:
```suggestion
/// Create a general line break segmenter with a word break option.
```

**Review by PgBiel — COMMENTED** [review]
This looks very useful! Here are some personal nitpicks; feel free to disagree (these are just my opinions).

File: crates/typst/src/text/linebreak.rs [hunk 4]
```diff
@@ -42,3 +44,26 @@ impl Behave for Packed<LinebreakElem> {
...
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Cast)]
+pub enum WordBreak {
+    /// Words break according to their customary rules.
+    Normal,
+
+    /// Breaking is allowed within “words”.
```
[the remark this answers is not in the document]
Comment (RanolP) [reply crates/typst/src/text/linebreak.rs]: I just copy-pasted from the CSS spec [(ref)](https://drafts.csswg.org/css-text/#word-break-property) there, and I also agree with your opinion. I'll remove the quotes. (e29cd741)

File: crates/typst/src/text/linebreak.rs [hunk 5]
```diff
@@ -42,3 +44,26 @@ impl Behave for Packed<LinebreakElem> {
         Behaviour::Destructive
     }
 }
+
+/// A word-break mode on text
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Cast)]
+pub enum WordBreak {
+    /// Words break according to their customary rules.
+    Normal,
```
[the remark this answers is not in the document]
Comment (RanolP) [reply crates/typst/src/text/linebreak.rs]: Good point. But it is a bit complicated to explain all the rules so I just wrote about the differences between English and Chinese/Japanese and there must be more customary rules for other languages. (6cdebe54).

File: crates/typst/src/layout/inline/linebreak.rs (hunk 3 above)
Comment (RanolP) [reply crates/typst/src/layout/inline/linebreak.rs]: Fixed in 84f11739

**Comment by RanolP** [comment]
Thanks for the review! The parts you pointed out are very reasonable to me, so I supplemented the documentation based on the comments.

**Comment by laurmaedje** [comment]
Thanks for the PR! We've been and are still busy preparing for the release of Typst 0.11 and aren't merging any new features anymore that require at list a bit of design and discussion until that has happened. I'll get back to this PR once the release is out.

**Comment by peng1999** [comment]
I suggest using table in test, to make it concise and compact.

Something like this:

```typ
#let k-text = [일부 텍스트]
#let e-text = [Some text]
#table(columns: (auto, 12em, 12em),
  table.cell(x: 1, y: 0)[한국어],
  table.cell(x: 2, y: 0)[English],
  ..for (i, p) in ("normal", "keep-all", "break-all").enumerate() {
    (
      table.cell(x: 0, y: i + 1, p),
      table.cell(x: 1, y: i + 1, {
        set text(word-break: p)
        k-text
      }),
      table.cell(x: 2, y: i + 1, {
        set text(word-break: p)
        e-text
      }),
    )
  }
)
```

**Comment by RanolP** [comment]
> I suggest using table in test, to make it concise and compact.
> 
> Something like this:

It is a good idea but because of the nature of keep-all - and lack of overflow-wrap support at this time -, the overlap happens (see 2nd line of keep-all in image). I don't know how to mitigate this. Please let me know if anyone knows how to do so.

image

Anyway, extracting texts using `let` and `for` to reduce repetition is good for maintenance. I'll refactor the test file.

**Comment by laurmaedje** [comment]
I am not a fan of the CSS terminology. If I read just `word-break: "break-all"`, I have no idea what it does. 

- `"normal"` seems to be the smart default, so it should probably be `auto`.
- `"keep-all"` could perhaps be `false`?
- `"break-all"` could maybe be `true`?

This is just one potential alternative. We should also consider whether there is other potential linebreak configuration that interacts with this?

**Comment by RanolP** [comment]
First of all, I just used the exact css term for less confusing with the css's one because the implementation(icu_segmenter crate) uses the css term. If it is not confusing to users, it's okay to be changed. but I'm not sure at this moment because I'm just Korean and don't know how other languages (especially east-Asian) speakers accept this option.
Next, since the Web is widely used around the world, it must contain the options meeting most requirements the people need. [CSS Text Module - Line Breaking](https://drafts.csswg.org/css-text-3/#line-breaking) is a good point to dig with for other potential linebreak config. I just found `overflow-wrap` (need to restructure line break algorithm of typst) and `line-break` (an option from icu_segmenter).

**Comment by laurmaedje** [comment]
I've discussed this with @reknih and we came to the conclusion that we'd like to flesh out a comprehensive solution rather than adding support for one specific CSS property. However, we ourselves don't currently have the time to do that. Maybe, we can just apply the property internally based on the language to fix #1164?

**Comment by RanolP** [comment]
Even though most Korean people want this feature, I wouldn't recommend that direction if there's no way to opt out of this behavior (some Korean publication styles don't use keep all and do a line wrap inside the word)—however, as you suggested, setting the default value for Korean looks good to me. I'd like to investigate this more and when I come up with a mature solution, I'll propose an RFC. At this time, I'll close this PR cuz there's no solution yet.
